[AArch64] Add support for range prefetch intrinsic

kmclaughlin-arm opened the pull request:
This patch adds support in Clang for the RPRFM instruction, by adding the following intrinsics:

```
void __pldx_range(unsigned int *access_kind*, unsigned int retention_policy,
                               signed int length*, unsigned int count, signed int stride,
                               size_t reuse distance,  void const *addr);

void __pld_range(unsigned int access_kind*, unsigned int retention_policy,
                              uint64_t metadata, void const *addr);
```

The `__ARM_PREFETCH_RANGE` macro can be used to test whether these intrinsics
are implemented. If the RPRFM instruction is not available, this instruction is a NOP.

This implements the following ACLE proposal:
https://github.com/ARM-software/acle/pull/423

kmclaughlin-arm committed:
[AArch64] Add support for range prefetch intrinsic

This patch adds support in Clang for the RPRFM instruction, which is
available when FEAT_RPRFM is defined:

  void __rpld(int64_t access_kind, uint64_t retention_policy
              uint64_t reuse distance, int64_t stride,
              uint64_t count, int64_t length, void const *addr);

If FEAT_RPRFM is not available, this instruction is a NOP.

This implements the following ACLE proposal:
https://github.com/ARM-software/acle/pull/423

kmclaughlin-arm committed:
- Change range of Count argument (1 to 65535) and subtract 1 during lowering
- Add Verifier tests
- Run clang-format

kmclaughlin-arm committed:
- Rename __rpld to __pldx_range
- Add second intrinsic (__pld_range) which takes a single metadata value
- Reorder the arguments of __pldx_range
- Removed RPRFM feature macro & added __ARM_PREFETCH_RANGE
- Map ReuseDistance to decreasing powers of two from 512MiB-32KiB

CarolineConcatto commented on clang/test/CodeGen/arm_acle.c:
```diff
@@ -164,6 +164,19 @@ void test_pld() {
+// AArch64-NEXT:    ret void
+//
+void test_rpld() {
+  __rpld(1, 1, 15, -2048, 65536, 2040, 0);
```
nit, maybe change access kind to 0, so we can check against retention policy, because both can be 0 or 1.

CarolineConcatto commented on llvm/test/CodeGen/AArch64/range-prefetch.ll:
```diff
@@ -0,0 +1,28 @@
+; CHECK-NEXT: mov   x9, #4192256
+; CHECK-NEXT: movk  x8, #2040
+; CHECK-NEXT: orr   x9, x9, #0x1fe0000000000
+; CHECK-NEXT: movk  x8, #65472, lsl #16
```
Why these lanes never change, even if you change the values of stride, count and length?

kmclaughlin-arm replied:
As I understand it, these instructions are moving the value representing the metadata into `x8` & `x9`, not the individual values for stride, count, length & distance. There are only two values being calculated because two of the intrinsics in this test are using identical values to the others.

I've rewritten this test a bit to try and make it easier to check that the correct values are being created here, especially for reuse distance.

CarolineConcatto reviewed (commented):
I believe more changes will come, due to changes in the ACLE proposal, but just in case I left some comments.

paulwalker-arm commented on llvm/include/llvm/IR/IntrinsicsAArch64.td:
```diff
@@ -76,6 +76,19 @@ def int_aarch64_prefetch : Intrinsic<[],
+     llvm_i32_ty, llvm_i64_ty],
+    [IntrInaccessibleMemOrArgMemOnly, IntrWillReturn, ReadOnly<ArgIndex<0>>,
+     ImmArg<ArgIndex<1>>, ImmArg<ArgIndex<2>>, ImmArg<ArgIndex<3>>, ImmArg<ArgIndex<4>>,
+     ImmArg<ArgIndex<5>>, ImmArg<ArgIndex<6>>]>;
```
I'm not against having such, but do we need a user friendly LLVM intrinsic? You could let clang do the encoding and keep the IR side simple.

kmclaughlin-arm committed:
- Fix range of length & stride arguments
- Move handling of immediate values to Clang
- Remove @llvm.aarch64.range.prefetch.imm intrinsic

kmclaughlin-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6701,6 +6701,38 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
+    Check(Length > -2049 && Length < 2041,
+          "length argument to llvm.aarch64.range.prefetch.imm must be -2048 - "
+          "2040",
+          Call);
```
These checks have been removed along with the `_imm` intrinsic, but I've fixed the ranges for stride & length.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -536,6 +536,7 @@ def SDT_AArch64trivec : SDTypeProfile<1, 3, [SDTCisVec<0>, SDTCisSameAs<0,1>,
                                            SDTCisSameAs<0,3>]>;
 def SDT_AArch64TCRET : SDTypeProfile<0, 2, [SDTCisPtrTy<0>]>;
 def SDT_AArch64PREFETCH : SDTypeProfile<0, 2, [SDTCisVT<0, i32>, SDTCisPtrTy<1>]>;
+def SDT_AArch64RANGE_PREFETCH: SDTypeProfile<0, 3, [SDTCisVT<0, i32>, SDTCisPtrTy<1>, SDTCisPtrTy<2>]>;
```
Should the metadata operand be `SDTCisVT<2, i64>` rather than `PtrTy`?  I assume the current form works because the two are synonymous, but best not to rely on that.

paulwalker-arm commented on llvm/test/CodeGen/AArch64/range-prefetch.ll:
```diff
@@ -0,0 +1,29 @@
+; CHECK-NEXT:    ret
+  call void @llvm.aarch64.range.prefetch(ptr %a, i32 0, i32 0, i64 576460752305520640)
+  ret void
+}
```
Does this test anything that's not already covered by `range_prefetch_metadata_accesses`?

kmclaughlin-arm committed:
- Replace use of TARGET_HEADER_BUILTIN with BUILTIN
- Change metadata operand of node to SDTCisVT<2, i64>
- Removed @range_prefetch_metadata_const test
- Metadata argument of __builtin_arm_range_prefetch is now uint64_t

kmclaughlin-arm commented on clang/include/clang/Basic/BuiltinsAArch64.def:
```diff
@@ -96,6 +96,10 @@ TARGET_BUILTIN(__builtin_arm_jcvt, "Zid", "nc", "v8.3a")
 
+// Range Prefetch
+TARGET_HEADER_BUILTIN(__builtin_arm_range_prefetch_x, "vvC*UiUiiUiiz", "n", ARMACLE_H, ALL_LANGUAGES, "")
+TARGET_HEADER_BUILTIN(__builtin_arm_range_prefetch, "vvC*UiUiULLi", "n", ARMACLE_H, ALL_LANGUAGES, "")
```
This is now `WUi` to reflect the change from `unsigned long` to `uint64_t`.

CarolineConcatto reviewed (approved):
Thank you Kerry, 
The work looks good to me.

jthackray reviewed (approved):
Thanks, Kerry. LGTM.
